### ISSUE TITLE
Interface to get safeArea for notched devices (display cutouts) such as iPhone XR?

### ISSUE THREAD

**Issue by SDLBugzilla** [libsdl-org/SDL#3243]
# This bug report was migrated from our old Bugzilla tracker.

**Reported in version:** HG 2.1
**Reported for operating system, platform:** Windows 10, iPhone/iPod touch

# Comments on the original bug report:

On 2019-04-19 17:52:22 +0000, Dan Ginsburg wrote:

> Notched devices such as the iPhone X/iPhone XR require applications to not render UI over the "safeArea" such as the notch.  This can be queried from the UIView usinng safeAreaInsets:
> 
> https://developer.apple.com/documentation/uikit/uiview/2891103-safeareainsets
> 
> It appears to me that Android has a similar API:
> 
> https://developer.android.com/guide/topics/display-cutout
> 
> Unity exposes this concept via a safeArea:
> 
> https://connect.unity.com/p/updating-your-gui-for-the-iphone-x-and-other-notched-devices
> 
> Has an interface been considered to add this to SDL?  For now I can workaround it using platform-specific APIs but it seems this is probably a feature that SDL needs for mobile devices.

**Comment by icculus** [comment]
We have SDL_GetDisplayUsableBounds() which could honor the SafeAreaInset, but my understanding is you can draw to the whole display otherwise, and this is just to tell the app not to put something critical (or touchable) where the notch is...so this might need a new API. Doesn't have to be iOS-specific, since sooner or later Android is going to do this too, I'm sure.

(Also the newest MacBook Pros also have a notch, so check if there's a macOS equivalent API now.)

**Comment by StarStarJ** [comment]
I have the same problem for android with fullscreen (where `SDL_GetDisplayUsableBounds` wont make sense from its documentation)
I think for `SDL_WINDOW_FULLSCREEN` it should automatically respect the notch, since in fullscreen apps you dont expect to suddenly break the app, because the notch is blocking critical areas.(also bcs it makes it easier to port games from other OS to android without too much java struggle)
I have to use SDL2 in windowed mode rn because of this, which alters the screen size correctly

(e.g. this text should start at the left and be fully visible)

![grafik](https://user-images.githubusercontent.com/77283933/145189102-f25de04d-4610-4671-b8d1-6785cc25601c.png)

**Comment by slouken** [comment]
Here's a good overview of the issues involved:
https://blog.felgo.com/cross-platform-app-development/notch-developer-guide-ios-android

**Comment by slouken** [comment]
Here are the relevant Unity APIs:
https://docs.unity3d.com/ScriptReference/Screen-safeArea.html
https://docs.unity3d.com/ScriptReference/Screen-cutouts.html

**Comment by slouken** [comment]
More detail on iOS safe area and insets:
https://blog.smartnsoft.com/layout-guide-margins-insets-and-safe-area-demystified-on-ios-10-11-d6e7246d7cb8

**Comment by slouken** [comment]
Here's how VLC handled the notch on the new MacBook Pro:
https://code.videolan.org/videolan/vlc/-/issues/26291

**Comment by DominusExult** [comment]
@slouken this instantly crashes Exult at 
```
void SDL_SetWindowSafeAreaInsets(SDL_Window *window, int left, int right, int top, int bottom)
{
    window->safe_inset_left = left;
```

**Comment by DominusExult** [comment]
@Dragon-Baroque or do you see what needs to be fixed on our side to work with this change?

**Comment by slouken** [comment]
> @slouken this instantly crashes Exult at
> 
> ```
> void SDL_SetWindowSafeAreaInsets(SDL_Window *window, int left, int right, int top, int bottom)
> {
>     window->safe_inset_left = left;
> ```
This is an internal call which should never be called with a NULL window. What's the stack trace?

**Comment by DominusExult** [comment]
```
#0	0x000000010306210c in SDL_SetWindowSafeAreaInsets at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/SDL_video.c:3834
#1	0x0000000102f4044c in -[SDL_uikitview safeAreaInsetsDidChange] at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/uikit/SDL_uikitview.m:375
#2	0x00000001858ce778 in -[UIView _safeAreaInsetsDidChangeFromOldInsets:] ()
#3	0x0000000184c8ca70 in -[UIViewController _updateViewSafeAreaInsetsAndEagerlyUpdateContentScrollView:] ()
#4	0x0000000184c8d3bc in -[UIViewController _setContentOverlayInsets:andLeftMargin:rightMargin:] ()
#5	0x0000000184c8ec24 in -[UIViewController _updateContentOverlayInsetsFromParentIfNecessary] ()
#6	0x0000000184c8d534 in -[UIViewController _updateContentOverlayInsetsForSelfAndChildren] ()
#7	0x0000000184c9a214 in -[UIViewController _updateLayoutForStatusBarAndInterfaceOrientation] ()
#8	0x0000000184c9e540 in -[UIViewController viewDidMoveToWindow:shouldAppearOrDisappear:] ()
#9	0x00000001858ee3bc in -[UIView(Internal) _didMoveFromWindow:toWindow:] ()
#10	0x00000001858ee200 in -[UIView(Internal) _didMoveFromWindow:toWindow:] ()
#11	0x00000001858e3a90 in __45-[UIView(Hierarchy) _postMovedFromSuperview:]_block_invoke ()
#12	0x00000001858e39a8 in -[UIView _postMovedFromSuperview:] ()
#13	0x00000001858f06b0 in -[UIView(Internal) _addSubview:positioned:relativeTo:] ()
#14	0x0000000184cb7090 in -[UIViewControllerBuiltinTransitionViewAnimator animateTransition:] ()
#15	0x0000000184cbd250 in ___UIViewControllerTransitioningRunCustomTransition_block_invoke_3 ()
#16	0x0000000184df4538 in +[UIKeyboardSceneDelegate _pinInputViewsForKeyboardSceneDelegate:onBehalfOfResponder:duringBlock:] ()
#17	0x0000000184cbd1e8 in ___UIViewControllerTransitioningRunCustomTransition_block_invoke_2 ()
#18	0x00000001858ea4f0 in +[UIView(Animation) _setAlongsideAnimations:toRunByEndOfBlock:] ()
#19	0x0000000184cbd088 in _UIViewControllerTransitioningRunCustomTransition ()
#20	0x0000000184bb1c30 in __56-[UIPresentationController runTransitionForCurrentState]_block_invoke_3 ()
#21	0x0000000184bb67f8 in +[UIPresentationController _scheduleTransition:] ()
#22	0x0000000184bb1384 in -[UIPresentationController runTransitionForCurrentState] ()
#23	0x0000000184baefc0 in -[UIPresentationController _presentWithAnimationController:inWindow:interactionController:target:didFinish:] ()
#24	0x0000000185406688 in -[UIWindow addRootViewControllerViewIfPossible] ()
#25	0x0000000185407654 in -[UIWindow setRootViewController:] ()
#26	0x0000000102f3f28c in -[SDL_uikitview setSDLWindow:] at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/uikit/SDL_uikitview.m:140
#27	0x000000010306f720 in UIKit_Metal_CreateView at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/uikit/SDL_uikitmetalview.m:108
#28	0x0000000103065538 in SDL_Metal_CreateView at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/SDL_video.c:5653
#29	0x0000000102ec901c in METAL_CreateRenderer at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/render/metal/SDL_render_metal.m:2007
#30	0x0000000102fa5018 in SDL_CreateRendererWithProperties at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/render/SDL_render.c:1024
#31	0x0000000102fa4c84 in SDL_CreateRenderer at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/render/SDL_render.c:1160
#32	0x0000000100c38eb4 in Image_window::create_scale_surfaces(int, int, int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/imagewin.cc:679
#33	0x0000000100c38c94 in Image_window::try_scaler(int, int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/imagewin.cc:839
#34	0x0000000100c378ec in Image_window::create_surface(unsigned int, unsigned int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/imagewin.cc:533
#35	0x0000000100823c5c in Image_window::Image_window(Image_buffer*, int, int, int, int, int, bool, int, Image_window::FillMode, int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/imagewin.h:338
#36	0x0000000100823af4 in Image_window8::Image_window8(unsigned int, unsigned int, unsigned int, unsigned int, int, bool, int, Image_window::FillMode, unsigned int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/iwin8.cc:61
#37	0x0000000100823cf0 in Image_window8::Image_window8(unsigned int, unsigned int, unsigned int, unsigned int, int, bool, int, Image_window::FillMode, unsigned int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/iwin8.cc:63
#38	0x0000000100b75008 in Game_window::Game_window(int, int, bool, int, int, int, int, Image_window::FillMode, unsigned int) at /Users/Dominus/Code/Git/exultsdl3new/gamewin.cc:317
#39	0x0000000100b76444 in Game_window::Game_window(int, int, bool, int, int, int, int, Image_window::FillMode, unsigned int) at /Users/Dominus/Code/Git/exultsdl3new/gamewin.cc:311
#40	0x0000000100c44cbc in setup_video(bool, int, int, int, int, int, int, int, Image_window::FillMode, int) at /Users/Dominus/Code/Git/exultsdl3new/exult.cc:3031
#41	0x0000000100c45d38 in Init() at /Users/Dominus/Code/Git/exultsdl3new/exult.cc:829
#42	0x0000000100c40ab8 in exult_main(char const*) at /Users/Dominus/Code/Git/exultsdl3new/exult.cc:608
#43	0x0000000100c3f30c in ::SDL_main(int, char **) at /Users/Dominus/Code/Git/exultsdl3new/exult.cc:421
#44	0x0000000102f814e4 in -[SDLUIKitDelegate postFinishLaunch] at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/uikit/SDL_uikitappdelegate.m:408
#45	0x0000000180dd41a8 in __NSFireDelayedPerform ()
#46	0x000000018040f548 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ ()
#47	0x000000018040f204 in __CFRunLoopDoTimer ()
#48	0x000000018040e8a0 in __CFRunLoopDoTimers ()
#49	0x0000000180408fec in __CFRunLoopRun ()
#50	0x00000001804084d4 in CFRunLoopRunSpecific ()
#51	0x000000018ef2aae4 in GSEventRunModal ()
#52	0x00000001853d0a28 in -[UIApplication _run] ()
#53	0x00000001853d46b0 in UIApplicationMain ()
#54	0x0000000102f7fc08 in SDL_RunApp at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/uikit/SDL_uikitappdelegate.m:70
#55	0x0000000100c3e588 in main at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/include/SDL3/SDL_main_impl.h:208
#56	0x0000000101675544 in start_sim ()
#57	0x00000001018760e0 in start ()
```
The Exult SDL3 fork resides at https://github.com/Dragon-Baroque/exult/tree/exult-sdl3 (until SDL3 stabilizes :))

**Comment by DominusExult** [comment]
This is for the current SDL3 with current Exult SDL3 fork (which had to adapt to other changes). Crash remains the same but some of the lines changed since https://github.com/libsdl-org/SDL/commit/457ca3995c9929cac3a3a617cdc4145d010d6a1c

```
#0	0x0000000105119e04 in SDL_SetWindowSafeAreaInsets at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/SDL_video.c:3847
#1	0x0000000104ff7c90 in -[SDL_uikitview safeAreaInsetsDidChange] at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/uikit/SDL_uikitview.m:380
#2	0x00000001858ce778 in -[UIView _safeAreaInsetsDidChangeFromOldInsets:] ()
#3	0x0000000184c8ca70 in -[UIViewController _updateViewSafeAreaInsetsAndEagerlyUpdateContentScrollView:] ()
#4	0x0000000184c8d3bc in -[UIViewController _setContentOverlayInsets:andLeftMargin:rightMargin:] ()
#5	0x0000000184c8ec24 in -[UIViewController _updateContentOverlayInsetsFromParentIfNecessary] ()
#6	0x0000000184c8d534 in -[UIViewController _updateContentOverlayInsetsForSelfAndChildren] ()
#7	0x0000000184c9a214 in -[UIViewController _updateLayoutForStatusBarAndInterfaceOrientation] ()
#8	0x0000000184c9e540 in -[UIViewController viewDidMoveToWindow:shouldAppearOrDisappear:] ()
#9	0x00000001858ee3bc in -[UIView(Internal) _didMoveFromWindow:toWindow:] ()
#10	0x00000001858ee200 in -[UIView(Internal) _didMoveFromWindow:toWindow:] ()
#11	0x00000001858e3a90 in __45-[UIView(Hierarchy) _postMovedFromSuperview:]_block_invoke ()
#12	0x00000001858e39a8 in -[UIView _postMovedFromSuperview:] ()
#13	0x00000001858f06b0 in -[UIView(Internal) _addSubview:positioned:relativeTo:] ()
#14	0x0000000184cb7090 in -[UIViewControllerBuiltinTransitionViewAnimator animateTransition:] ()
#15	0x0000000184cbd250 in ___UIViewControllerTransitioningRunCustomTransition_block_invoke_3 ()
#16	0x0000000184df4538 in +[UIKeyboardSceneDelegate _pinInputViewsForKeyboardSceneDelegate:onBehalfOfResponder:duringBlock:] ()
#17	0x0000000184cbd1e8 in ___UIViewControllerTransitioningRunCustomTransition_block_invoke_2 ()
#18	0x00000001858ea4f0 in +[UIView(Animation) _setAlongsideAnimations:toRunByEndOfBlock:] ()
#19	0x0000000184cbd088 in _UIViewControllerTransitioningRunCustomTransition ()
#20	0x0000000184bb1c30 in __56-[UIPresentationController runTransitionForCurrentState]_block_invoke_3 ()
#21	0x0000000184bb67f8 in +[UIPresentationController _scheduleTransition:] ()
#22	0x0000000184bb1384 in -[UIPresentationController runTransitionForCurrentState] ()
#23	0x0000000184baefc0 in -[UIPresentationController _presentWithAnimationController:inWindow:interactionController:target:didFinish:] ()
#24	0x0000000185406688 in -[UIWindow addRootViewControllerViewIfPossible] ()
#25	0x0000000185407654 in -[UIWindow setRootViewController:] ()
#26	0x0000000104ff6aac in -[SDL_uikitview setSDLWindow:] at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/uikit/SDL_uikitview.m:140
#27	0x0000000105127498 in UIKit_Metal_CreateView at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/uikit/SDL_uikitmetalview.m:115
#28	0x000000010511d230 in SDL_Metal_CreateView at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/SDL_video.c:5666
#29	0x0000000104f8085c in METAL_CreateRenderer at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/render/metal/SDL_render_metal.m:2007
#30	0x000000010505c8c0 in SDL_CreateRendererWithProperties at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/render/SDL_render.c:1028
#31	0x000000010505c530 in SDL_CreateRenderer at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/render/SDL_render.c:1166
#32	0x0000000102c98e60 in Image_window::create_scale_surfaces(int, int, int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/imagewin.cc:679
#33	0x0000000102c98c40 in Image_window::try_scaler(int, int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/imagewin.cc:839
#34	0x0000000102c97898 in Image_window::create_surface(unsigned int, unsigned int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/imagewin.cc:534
#35	0x0000000102883bd4 in Image_window::Image_window(Image_buffer*, int, int, int, int, int, bool, int, Image_window::FillMode, int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/imagewin.h:338
#36	0x0000000102883a6c in Image_window8::Image_window8(unsigned int, unsigned int, unsigned int, unsigned int, int, bool, int, Image_window::FillMode, unsigned int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/iwin8.cc:61
#37	0x0000000102883c68 in Image_window8::Image_window8(unsigned int, unsigned int, unsigned int, unsigned int, int, bool, int, Image_window::FillMode, unsigned int) at /Users/Dominus/Code/Git/exultsdl3new/imagewin/iwin8.cc:63
#38	0x0000000102bd4f9c in Game_window::Game_window(int, int, bool, int, int, int, int, Image_window::FillMode, unsigned int) at /Users/Dominus/Code/Git/exultsdl3new/gamewin.cc:317
#39	0x0000000102bd63d8 in Game_window::Game_window(int, int, bool, int, int, int, int, Image_window::FillMode, unsigned int) at /Users/Dominus/Code/Git/exultsdl3new/gamewin.cc:311
#40	0x0000000102ca4c68 in setup_video(bool, int, int, int, int, int, int, int, Image_window::FillMode, int) at /Users/Dominus/Code/Git/exultsdl3new/exult.cc:3039
#41	0x0000000102ca5cec in Init() at /Users/Dominus/Code/Git/exultsdl3new/exult.cc:830
#42	0x0000000102ca0a64 in exult_main(char const*) at /Users/Dominus/Code/Git/exultsdl3new/exult.cc:608
#43	0x0000000102c9f2b8 in ::SDL_main(int, char **) at /Users/Dominus/Code/Git/exultsdl3new/exult.cc:421
#44	0x0000000105038d24 in -[SDLUIKitDelegate postFinishLaunch] at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/uikit/SDL_uikitappdelegate.m:408
#45	0x0000000180dd41a8 in __NSFireDelayedPerform ()
#46	0x000000018040f548 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ ()
#47	0x000000018040f204 in __CFRunLoopDoTimer ()
#48	0x000000018040e8a0 in __CFRunLoopDoTimers ()
#49	0x0000000180408fec in __CFRunLoopRun ()
#50	0x00000001804084d4 in CFRunLoopRunSpecific ()
#51	0x000000018ef2aae4 in GSEventRunModal ()
#52	0x00000001853d0a28 in -[UIApplication _run] ()
#53	0x00000001853d46b0 in UIApplicationMain ()
#54	0x0000000105037448 in SDL_RunApp at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/src/video/uikit/SDL_uikitappdelegate.m:70
#55	0x0000000102c9e534 in main at /Users/Dominus/Code/Git/exultsdl3new/ios/SDL/include/SDL3/SDL_main_impl.h:208
#56	0x00000001036d5544 in start_sim ()
#57	0x000000010377a0e0 in start ()

```

**Comment by slouken** [comment]
Yep, I know what the problem is, I'm working on a fix.

**Comment by slouken** [comment]
Fixed in 45ecea134, thanks!

**Comment by slouken** [comment]
You're welcome! You guys are willing to stay on the bleeding edge and give us valuable bug reports, we really appreciate that!
